sway-ir/src/verify.rs

```diff
@@ -1017,7 +1017,7 @@ impl<'a, 'eng> InstructionVerifier<'a, 'eng> {
     // This is a really common operation above... calling `Value::get_type()` and then failing when
     // two don't match.
     fn opt_ty_not_eq(&self, l_ty: &Option<Type>, r_ty: &Option<Type>) -> bool {
-        l_ty.is_none() || r_ty.is_none() || !l_ty.unwrap().eq(self.context, r_ty.as_ref().unwrap())
+        l_ty.is_none() || r_ty.is_none() && !l_ty.unwrap().eq(self.context, r_ty.as_ref().unwrap())
```

Comment:
Priority of operators will lead to this which is I assume not intended:
```
l_ty.is_none() || ( r_ty.is_none() && !l_ty.unwrap().eq(self.context, r_ty.as_ref().unwrap()) )
```
If `l_ty` is `Some` and `r_ty` is `None` the last operand will be executed leading to unwrapping of `r_ty` which is `None`.